In [1]:
!pip install -U deep-translator

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [21]:
import pandas as pd
import numpy as np
import re
import string
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from google.colab import files
from deep_translator import GoogleTranslator

In [3]:
df_train = pd.read_csv("https://adamhyman-public.s3.amazonaws.com/w266/from_kaggle/train.csv")
df_test = pd.read_csv("https://adamhyman-public.s3.amazonaws.com/w266/from_kaggle/test.csv")
df_test_labels = pd.read_csv("https://adamhyman-public.s3.amazonaws.com/w266/from_kaggle/test_labels.csv")

print (df_train.shape)
print (df_test.shape)
print (df_test_labels.shape)


(159571, 8)
(153164, 2)
(153164, 7)


In [4]:
display (df_train.head(5))
display (df_test.head(5))
display (df_test_labels.head(5))

id                                       comment_text  toxic  \
0  0000997932d777bf  Explanation\nWhy the edits made under my usern...      0   
1  000103f0d9cfb60f  D'aww! He matches this background colour I'm s...      0   
2  000113f07ec002fd  Hey man, I'm really not trying to edit war. It...      0   
3  0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...      0   
4  0001d958c54c6e35  You, sir, are my hero. Any chance you remember...      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       0              0  
3             0        0       0       0              0  
4             0        0       0       0              0

id                                       comment_text
0  00001cee341fdb12  Yo bitch Ja Rule is more succesful then you'll...
1  0000247867823ef7  == From RfC == \n\n The title is fine as it is...
2  00013b17ad220c46  " \n\n == Sources == \n\n * Zawe Ashton on Lap...
3  00017563c3f7919a  :If you have a look back at the source, the in...
4  00017695ad8997eb          I don't anonymously edit articles at all.

id  toxic  severe_toxic  obscene  threat  insult  \
0  00001cee341fdb12     -1            -1       -1      -1      -1   
1  0000247867823ef7     -1            -1       -1      -1      -1   
2  00013b17ad220c46     -1            -1       -1      -1      -1   
3  00017563c3f7919a     -1            -1       -1      -1      -1   
4  00017695ad8997eb     -1            -1       -1      -1      -1   

   identity_hate  
0             -1  
1             -1  
2             -1  
3             -1  
4             -1

Join Test Labels (contains the labels) to Test (contains the comment)

In [5]:

df_test = df_test.merge(df_test_labels, on='id', how='left')

print (df_test.shape)
display (df_test.head(5))

(153164, 8)


id                                       comment_text  toxic  \
0  00001cee341fdb12  Yo bitch Ja Rule is more succesful then you'll...     -1   
1  0000247867823ef7  == From RfC == \n\n The title is fine as it is...     -1   
2  00013b17ad220c46  " \n\n == Sources == \n\n * Zawe Ashton on Lap...     -1   
3  00017563c3f7919a  :If you have a look back at the source, the in...     -1   
4  00017695ad8997eb          I don't anonymously edit articles at all.     -1   

   severe_toxic  obscene  threat  insult  identity_hate  
0            -1       -1      -1      -1             -1  
1            -1       -1      -1      -1             -1  
2            -1       -1      -1      -1             -1  
3            -1       -1      -1      -1             -1  
4            -1       -1      -1      -1             -1

Get rid of records from Test that aren't labelled.

In [6]:
df_test = df_test[df_test.toxic != -1]
df_test = df_test.reset_index(drop=True)

print (df_test.shape)
display (df_test.head(20))

(63978, 8)


id                                       comment_text  \
0   0001ea8717f6de06  Thank you for understanding. I think very high...   
1   000247e83dcc1211                   :Dear god this site is horrible.   
2   0002f87b16116a7f  "::: Somebody will invariably try to add Relig...   
3   0003e1cccfd5a40a  " \n\n It says it right there that it IS a typ...   
4   00059ace3e3e9a53  " \n\n == Before adding a new product to the l...   
5   000663aff0fffc80                           this other one from 1897   
6   000689dd34e20979  == Reason for banning throwing == \n\n This ar...   
7   000844b52dee5f3f             |blocked]] from editing Wikipedia.   |   
8   00091c35fa9d0465  == Arabs are committing genocide in Iraq, but ...   
9   000968ce11f5ee34  Please stop. If you continue to vandalize Wiki...   
10  0009734200a85047  == Energy  == \n\n I have edited the introduct...   
11  000a02d807ae0254  @RedSlash, cut it short. If you have sources s...   
12  000bafe2080bba82  . \n\n           Jews are not a race because y...   
13  000bf0a9894b2807  :::If Ollie or others think that one list of t...   
14  000c9b92318552d1               Professors to the Manhatten Project.   
15  000cf60dbaed8c02  " \n\n :Not sure whether this is notable enoug...   
16  000d4f120d5a7303                       일이삼사오육칠팔구하고십이요 에헤헤 으헤 으헤 으허허   
17  001068b809feee6b  " \n\n ==balance== \n This page has one senten...   
18  0011cefc680993ba                      REDIRECT Talk:Mi Vida Eres Tú   
19  0011ef6aa33d42e6  " \n I'm not convinced that he was blind. Wher...   

    toxic  severe_toxic  obscene  threat  insult  identity_hate  
0       0             0        0       0       0              0  
1       0             0        0       0       0              0  
2       0             0        0       0       0              0  
3       0             0        0       0       0              0  
4       0             0        0       0       0              0  
5       0             0        0       0       0              0  
6       0             0        0       0       0              0  
7       0             0        0       0       0              0  
8       1             0        0       0       0              0  
9       0             0        0       0       0              0  
10      0             0        0       0       0              0  
11      0             0        0       0       0              0  
12      0             0        0       0       0              0  
13      0             0        0       0       0              0  
14      0             0        0       0       0              0  
15      0             0        0       0       0              0  
16      0             0        0       0       0              0  
17      0             0        0       0       0              0  
18      0             0        0       0       0              0  
19      0             0        0       0       0              0

In [ ]:

print ('Train')
print ('Train shape:  ' + str(df_train.shape))
print ('Toxic:  ' + str(len(df_train[df_train.toxic == 1])))
print ('Severe Toxic:  ' + str(len(df_train[df_train.severe_toxic == 1])))
print ('Obscene:  ' + str(len(df_train[df_train.obscene == 1])))
print ('Threat:  ' + str(len(df_train[df_train.threat == 1])))
print ('Insult:  ' + str(len(df_train[df_train.insult == 1])))
print ('Identity Hate:  ' + str(len(df_train[df_train.identity_hate == 1])))


print ('\n\nTest')
print ('Test shape:  ' + str(df_test.shape))
print ('Toxic:  ' + str(len(df_test[df_test.toxic == 1])))
print ('Severe Toxic:  ' + str(len(df_test[df_test.severe_toxic == 1])))
print ('Obscene:  ' + str(len(df_test[df_test.obscene == 1])))
print ('Threat:  ' + str(len(df_test[df_test.threat == 1])))
print ('Insult:  ' + str(len(df_test[df_test.insult == 1])))
print ('Identity Hate:  ' + str(len(df_test[df_test.identity_hate == 1])))




Train
Train shape:  (159571, 8)
Toxic:  15294
Severe Toxic:  1595
Obscene:  8449
Threat:  478
Insult:  7877
Identity Hate:  1405


Test
Test shape:  (63978, 8)
Toxic:  6090
Severe Toxic:  367
Obscene:  3691
Threat:  211
Insult:  3427
Identity Hate:  712


Delete records with missing comments

In [7]:
df_test = df_test[df_test["comment_text"] != '']



Get rid of records which have 1 count, based on the columns we are stratifying by.

In [8]:

types_of_toxicity = ['toxic', 'severe_toxic', 'obscene', 'threat', 'insult', 'identity_hate']

train_counts = df_train.groupby(types_of_toxicity)['id'].count().reset_index(name='counts')
train_counts_gt_one = train_counts[train_counts['counts'] > 1]


df_train = df_train.merge(train_counts_gt_one, on=types_of_toxicity, how='inner')
df_train = df_train.drop('counts', axis=1)

print (df_train.shape)
display (df_train)

(159569, 8)


id                                       comment_text  \
0       0000997932d777bf  Explanation\nWhy the edits made under my usern...   
1       000103f0d9cfb60f  D'aww! He matches this background colour I'm s...   
2       000113f07ec002fd  Hey man, I'm really not trying to edit war. It...   
3       0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...   
4       0001d958c54c6e35  You, sir, are my hero. Any chance you remember...   
...                  ...                                                ...   
159564  b69dae3e8c38fb63  wtf? \n\nwtf is your problem? why are you bein...   
159565  1d0abba33f555b04  "\n\n Treivas, Miller: you think you're safe o...   
159566  c797b58f2390ea5a  "::::niggardly is a perennial favourite in tha...   
159567  e2a35ba86bacfac8  "\n\n A beating \n\nHello Shovon, I see you've...   
159568  586cce763d6ee7b7  Jews for Jesus \nSome Jews love Jesus in a way...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159564      0             0        1       1       1              0  
159565      0             0        1       1       1              0  
159566      0             0        1       0       0              1  
159567      0             0        1       0       0              1  
159568      0             0        1       0       0              1  

[159569 rows x 8 columns]

# Add Toxigen for identity hate

In [9]:
!pip install datasets

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.7/468.7 KB 26.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 KB 10.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 71.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.9/132.9 KB 15.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 200.1/200.1 KB 20.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.2/212.2 KB 23.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 264.6/264.6 KB 30.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 158.8/158.8 KB 19.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.2/114.2 KB 12.5 MB/s eta 0:00:00


In [10]:
add_toxigen_for_identity_hate = True

In [13]:
if add_toxigen_for_identity_hate:
  
  from datasets import load_dataset

  access_token = "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"

  TG_data = load_dataset("skg/toxigen-data", name="train", use_auth_token=access_token) # 250k training examples
  #TG_annotations = load_dataset("skg/toxigen-data", name="annotated", use_auth_token=True) # Human study

  TG = pd.DataFrame(TG_data["train"]) # Toxigen is stored in the train split of a huggingface dataset
  TG = TG[TG.prompt_label == 1.0] # Keeps only records that are toxic
  display (TG.head())

  0%|          | 0/1 [00:00<?, ?it/s]

prompt  \
0  - the vast majority of african-americans are p...   
1  - black people from the middle ages have alway...   
2  - many african americans are so savage they lo...   
3  - the vast majority of african-americans are p...   
4  - whether a scientist or janitor, black people...   

                                          generation generation_method  group  \
0  if youre white, you should be proud to be whit...             ALICE  black   
1  the best thing that could have happened to the...             ALICE  black   
2  theres no hope for the future because theres n...             ALICE  black   
3         if youre white, you should be proud of it.             ALICE  black   
4  no matter how much money or power you have, if...             ALICE  black   

   prompt_label  roberta_prediction  
0             1               0.001  
1             1               0.044  
2             1               0.014  
3             1               0.002  
4             1               0.328

In [14]:
if add_toxigen_for_identity_hate:
  
  print ('Generation Method:  ALICE')
  for i in range(2):
    print ('Prompt:  ', end='')
    print (TG[TG['generation_method'] == 'ALICE'].iloc[i]['prompt'])
    print ('Generation:  ', end='')
    print (TG[TG['generation_method'] == 'ALICE'].iloc[i]['generation'])
    print ('\n', end = '')


  print ('Generation Method:  top-k')
  for i in range(2):
    print ('Prompt:  ', end='')
    print (TG[TG['generation_method'] == 'top-k'].iloc[i]['prompt'])
    print ('Generation:  ', end='')
    print (TG[TG['generation_method'] == 'top-k'].iloc[i]['generation'])
    print ('\n', end = '')

Generation Method:  ALICE
Prompt:  - the vast majority of african-americans are poor because they are stupid and they spend all their money on drugs and other unproductive things\n- naturally less productive than their caucasian counterparts, african-americans have become vastly important as menial workers\n- african-americans shouldnt be allowed to vote or drive because theyre too stupid to give a valid opinion\n- black people from the middle ages have always been watermelon-eating, musically-inclined, simple-minded brutes\n- many african americans are so savage they look like they should be hanging from trees eating watermelons as a food staple\n- blacks have been programmed to love watermelon. they submit to stronger races and crave authority like dogs obeying their owners commands\n-
Generation:  if youre white, you should be proud to be white, and you should be proud of your heritage.

Prompt:  - black people from the middle ages have always been watermelon-eating, musically-incli

In [15]:
if add_toxigen_for_identity_hate:
  print (len(TG['generation']))
  print (TG.groupby(['generation_method'])['generation_method'].count())


125672
generation_method
ALICE      4150
top-k    121522
Name: generation_method, dtype: int64


In [16]:
if add_toxigen_for_identity_hate:
  
  alice_df = TG[TG['generation_method']=='ALICE'].sample(n=320, random_state=1)
  top_k_df = TG[TG['generation_method']=='top-k'].sample(n=320, random_state=1)
  identity_hate_df = pd.concat([top_k_df, alice_df])

  identity_hate_df.reset_index(drop=True, inplace=True)

  print (identity_hate_df.columns)


  def  fix_new_lines(text):
      text = re.sub(r"\\n", ". ", text)
      text = re.sub(r"\n-", ". ", text)
      return text

  for i in range(4):
    print (identity_hate_df['prompt'][i])

  identity_hate_df["prompt"] = identity_hate_df["prompt"].apply(lambda text: fix_new_lines(text))

  print ('-----------------------------------------------')

  for i in range(4):
    print (identity_hate_df['prompt'][i])


  identity_hate_df = identity_hate_df[['prompt']]
  identity_hate_df.rename(columns={"prompt": "comment_text"}, inplace=True)

  display (identity_hate_df.head(5))

Index(['prompt', 'generation', 'generation_method', 'group', 'prompt_label',
       'roberta_prediction'],
      dtype='object')
- if their countries were really so great, why are the latino people flocking to the border and begging to get in?\n- uneducated hispanics are lazy people who would rather be welfare recipients then workers\n- terrible diseases are caused and spead by latinos. wasps and fleas were introduced to the US by the hispanic people\n- when it comes down to it, latinos are mostly good for being dirty farmhands who smuggle drugs across the border\n- immigration is being used as a political weapon. the latinos riot and demand entry as a bargaining chip for whatever political reason they might have\n- it is more cost efficient for hispanics to live here then it is for them to live in their own countries\n-
- women should have jobs but only as receptionists, nurses, or teachers\n- women alreayd have jobs, they are housewives\n- if she has kids, try not to hire her because

comment_text
0  - if their countries were really so great, why...
1  - women should have jobs but only as reception...
2  - are middle easterners really people?. - i wa...
3  - most of them do not deserve the bad rep they...
4  - it is more cost efficient for hispanics to l...

In [17]:
if add_toxigen_for_identity_hate:

  #  Add additional columns and join to df_train

  identity_hate_df['id'] = 'toxigen'
  identity_hate_df['toxic'] = 1
  identity_hate_df['severe_toxic'] = 0
  identity_hate_df['obscene'] = 0
  identity_hate_df['threat'] = 0
  identity_hate_df['insult'] = 0
  identity_hate_df['identity_hate'] = 1

  identity_hate_df = identity_hate_df[['id', 'comment_text', 'toxic', 'severe_toxic', 'obscene', 'threat', 'insult', 'identity_hate']]

  display (identity_hate_df.head(5))

id                                       comment_text  toxic  \
0  toxigen  - if their countries were really so great, why...      1   
1  toxigen  - women should have jobs but only as reception...      1   
2  toxigen  - are middle easterners really people?. - i wa...      1   
3  toxigen  - most of them do not deserve the bad rep they...      1   
4  toxigen  - it is more cost efficient for hispanics to l...      1   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              1  
1             0        0       0       0              1  
2             0        0       0       0              1  
3             0        0       0       0              1  
4             0        0       0       0              1

In [18]:
if add_toxigen_for_identity_hate:
  print ('df_train:  ' + str(df_train.shape[0]))
  print ('identity_hate_df:  ' + str(identity_hate_df.shape[0]))
  df_train = df_train.append(identity_hate_df, ignore_index = True)
  print ('df_train:  ' + str(df_train.shape[0]))

df_train:  159569
identity_hate_df:  640
df_train:  160209


<ipython-input-18-3d375cf3b331>:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_train = df_train.append(identity_hate_df, ignore_index = True)


# Add Back Translation

In [19]:
add_back_translation = True


In [22]:
english_comment = df_train['comment_text'][5]
display (english_comment)
chinese_comment = GoogleTranslator(source='en', target='zh-CN').translate(english_comment) 
display (chinese_comment)
english_comment_back = GoogleTranslator(source='zh-CN', target='en').translate(chinese_comment)
display (english_comment_back)

'"\n\nCongratulations from me as well, use the tools well. \xa0· talk "'

'“\n\n我也恭喜你，用好工具。 · 讲话 ”'

'"\n\nI congratulate you too, good tool. · Speech"'

In [23]:

def back_translate_chinese (english_comment):
  try:
    chinese_comment = GoogleTranslator(source='en', target='zh-CN').translate(english_comment) 
    english_comment_back = GoogleTranslator(source='zh-CN', target='en').translate(chinese_comment) 
    return english_comment_back
  except:
    return english_comment

def back_translate_french (english_comment):
  try:
    french_comment = GoogleTranslator(source='en', target='fr').translate(english_comment) 
    english_comment_back = GoogleTranslator(source='fr', target='en').translate(french_comment)
    return english_comment_back
  except:
    return english_comment

def back_translate_hindi (english_comment):
  try:
    hindi_comment = GoogleTranslator(source='en', target='hi').translate(english_comment)
    english_comment_back = GoogleTranslator(source='hi', target='en').translate(hindi_comment)
    return english_comment_back
  except:
    return english_comment


for i in range(3):
  print (df_train['comment_text'][i])
  print (back_translate_chinese(df_train['comment_text'][i]))
  print (back_translate_french(df_train['comment_text'][i]))
  print (back_translate_hindi(df_train['comment_text'][i]))
  print ('\n')


Explanation
Why the edits made under my username Hardcore Metallica Fan were reverted? They weren't vandalisms, just closure on some GAs after I voted at New York Dolls FAC. And please don't remove the template from the talk page since I'm retired now.89.205.38.27
explain
Why are edits made under my username Hardcore Metallica Fan being reverted? They didn't mean to sabotage, they just closed some GA after I voted at the New York Dolls FAC. And please don't remove the template from the talk page as I'm retired now. 89.205.38.27
Explanation
Why have changes made under my Hardcore Metallica Fan username been reverted? It wasn't acts of vandalism, just the closing of some GAs after voting in New York Dolls FAC. And please don't delete the template from the talk page since I'm now retired.89.205.38.27
Explanation
Why were edits made under my username Hardcore Metallica Fan reverted? They weren't sabotage, just some GA shut down after I voted in the New York Dolls FAC. And please don't remo

In [24]:

print (len(df_train[(df_train['threat'] == 1) | (df_train['severe_toxic'] == 1)]))
print (len(df_train[df_train['threat'] == 1]))
print (len(df_train[df_train['severe_toxic'] == 1]))

1959
476
1593


## Creating the datasets for back translation

Threat has 476 records, we will back translate all of them through all three langagues.

Severe_Toxic has 1595 records.  We will take 500 of the records that are not included in threat, and back translate them through 1 langague

In [25]:
if add_back_translation:
  
  threat1 = df_train[(df_train['threat'] == 1)].copy()
  threat2 = threat1.copy()
  threat3 = threat1.copy()

  #  Get all severe toxic that are not threat
  severe = df_train[(df_train['threat'] == 0) & (df_train['severe_toxic'] == 1)].copy()
  print ('Severe only:  ' + str(len(severe)))

  #  Only take 400
  severe = severe.sample(n=500, random_state=1)

  print ('threat1:  ' + str(threat1.shape[0]))
  print ('severe:  ' + str(severe.shape[0]))

  display (severe.head(3))

Severe only:  1483
threat1:  476
severe:  500


id                                       comment_text  \
143547  48c1f8a8574a6909  "\n\n You fagg0t \n\nI was fuck!ng talking to ...   
143736  a6a606fff2c73afe  what i mean to say is that im a big fucknig id...   
143437  24abb6eaea97f43d  Whatever \n\nFucking block me then you punk as...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
143547      1             1        1       0       1              0  
143736      1             1        1       0       1              0  
143437      1             1        1       0       1              0

In [ ]:
#  If you need to split up the data and do back-translation in multiple notebooks

# threat2.to_csv('threat2.csv', index=False)
# files.download('threat2.csv')

# severe.to_csv('severe2.csv', index=False)
# files.download('severe2.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [26]:
if add_back_translation:
  
  # Threat

  threat1["comment_text"] = threat1["comment_text"][:499].apply(lambda text: back_translate_chinese(text))
  threat2["comment_text"] = threat2["comment_text"][:499].apply(lambda text: back_translate_french(text))
  threat3["comment_text"] = threat3["comment_text"][:499].apply(lambda text: back_translate_hindi(text))
  severe["comment_text"] = severe["comment_text"][:499].apply(lambda text: back_translate_chinese(text))
  

  print ('df_train:  ' + str(df_train.shape[0]))
  print ('threat1:  ' + str(threat1.shape[0]))
  print ('threat2:  ' + str(threat2.shape[0]))
  print ('threat3:  ' + str(threat3.shape[0]))
  print ('severe:  ' + str(severe.shape[0]))
  df_train = pd.concat([df_train, threat1, threat2, threat3, severe])
  print ('df_train:  ' + str(df_train.shape[0]))

df_train:  160209
threat1:  476
threat2:  476
threat3:  476
severe:  500
df_train:  162137


In [28]:
#  Check that back translation worked.

if add_back_translation:
  display (threat1.sample(n=10, random_state=1))

  display (threat1[threat1['id']=='7bbb41cbe7d745ec'])
  print (severe.iloc[0]['comment_text'])

id                                       comment_text  \
159531  2143adb07dec6c3f  (http://wiki.uschallengecoins.com/Main_Page)\n...   
159253  a8ffb7ff2cb8bb88  That's proper talk, you fucking idiot. I'll co...   
156208  7bbb41cbe7d745ec       Stop editing Karlie's page before I kill you   
157524  7f4d121e83056717  8=====D- - -\n8=====D- - - -\n8======D- - -\n8...   
156267  bc632baccddd180a  I don't care you damn well, like alexie and lu...   
158417  ef380323a975611c  Yadi\n\nfuck\nDon't you like YARDIE?\n\nUm\n\n...   
156242  e90346bf9e09529e  Fool!\n3RR\nPlease don't repeatedly undo other...   
159535  a62dfd807b61b58f  I will come back whenever I want. Then I'll hi...   
157509  14cf27f9495d131f  Why the fuck are you editing my contribution. ...   
159394  40d6431e631e36ac  That may be the case, but the exclusionary nat...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
159531      1             0        1       1       0              0  
159253      1             1        1       1       1              0  
156208      1             0        0       1       0              0  
157524      1             0        1       1       1              1  
156267      1             0        0       1       0              0  
158417      1             0        1       1       1              0  
156242      1             0        0       1       0              0  
159535      1             0        1       1       0              0  
157509      1             0        1       1       1              1  
159394      0             0        0       1       0              0

id                                  comment_text  toxic  \
156208  7bbb41cbe7d745ec  Stop editing Karlie's page before I kill you      1   

        severe_toxic  obscene  threat  insult  identity_hate  
156208             0        0       1       0              0

"

 fuck you

I'm fucking talking to you, you bitch. "Oh no, fuck you! Wreck me the fuck! Damn page I better cry the fuck out! ng wiki-river" "You father of vampires. suck my dick! ck. That was the last time I tried to be polite to someone. You ruined my faith in Wikipedia and I will now get my $200 donation back, you poor cvnt. "


Now that training data has been supplemented, we can clean it.

In [29]:
def  clean_text(text):
  try:
    text =  text.lower()
    text = re.sub(r"i'm", "i am", text)
#    text = re.sub(r"r", "", text)
    text = re.sub(r"he's", "he is", text)
    text = re.sub(r"she's", "she is", text)
    text = re.sub(r"it's", "it is", text)
    text = re.sub(r"that's", "that is", text)
    text = re.sub(r"what's", "that is", text)
    text = re.sub(r"where's", "where is", text)
    text = re.sub(r"how's", "how is", text)
    text = re.sub(r"'ll", " will", text)
    text = re.sub(r"'ve", " have", text)
    text = re.sub(r"'re", " are", text)
    text = re.sub(r"'d", " would", text)
    text = re.sub(r"'re", " are", text)
    text = re.sub(r"won't", "will not", text)
    text = re.sub(r"can't", "cannot", text)
    text = re.sub(r"n't", " not", text)
    text = re.sub(r"n'", "ng", text)
    text = re.sub(r"'bout", "about", text)
    text = re.sub(r"'til", "until", text)
    text = re.sub(r"[-()\"\#/@;:{}`+=~|!?,]", "", text)
    text = text.translate(str.maketrans('', '', string.punctuation)) 
    text = re.sub("(\W)"," ",text) 
  except:
    return text
   # text = re.sub('S*dS*s*','', text)
  return text

df_test["comment_text"] = df_test['comment_text'].apply(lambda text: clean_text(text))
df_train["comment_text"] = df_train['comment_text'].apply(lambda text: clean_text(text))

Stopping point to save and reload data, incase of disconnecint from Colab and due to API limitations on the backwards translation.

In [30]:
#  Save data

df_train.to_csv('train_data_stopping_point.csv', index=False)
files.download('train_data_stopping_point.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [33]:
#  Reload data

df_train = pd.read_csv("train_data_stopping_point_balanced.csv")
print ('Len after importing:  ' + str(len(df_train)))

df_train = df_train[df_train['toxic'].notna()]
print ('Len after dropping NAs:  ' + str(len(df_train)))

df_train = df_train.astype({'toxic':'int', 'severe_toxic':'int', 'obscene':'int', 'threat':'int', 'insult':'int', 'identity_hate':'int'})
display (df_train.head(5))
display (df_train[df_train['id']=='1ea25596e83d2042'])

Len after importing:  16430
Len after dropping NAs:  16414


id                                       comment_text  toxic  \
0  fa251ec9317c96e8  january 2015 utc  it is taken care of thank yo...      0   
1  673c22a421fcbe0e  your assumption is wrong levkamensky is a frie...      0   
2  4bc32763cdaeaee9  get sources first then add content you should ...      0   
3  4a00f665175362e3     why is my article bad   hi gilo  why did yo...      0   
4  db4af51f6f4728c2     should robespierre be classified as a liber...      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       0              0  
3             0        0       0       0              0  
4             0        0       0       0              0

id                                       comment_text  \
12502  1ea25596e83d2042  gaaaaaaaaaaaaaaaaaaaaaaaaay u r fucking gaey  ...   

       toxic  severe_toxic  obscene  threat  insult  identity_hate  
12502      1             0        1       0       1              1

In [34]:


threat_and_severe_toxic = df_train[(df_train['threat'] == 1) | (df_train['severe_toxic'] == 1)].copy()
print (threat_and_severe_toxic.groupby(['threat', 'severe_toxic'])['threat'].count())
print ('')

print ('Total records:  ' + str(len(df_train)))

print (df_train.groupby(['severe_toxic'])['severe_toxic'].count())
print ('')
print (df_train.groupby(['threat'])['threat'].count())
print ('')
print (df_train.groupby(['toxic'])['toxic'].count())
print ('')
print (df_train.groupby(['obscene'])['obscene'].count())
print ('')
print (df_train.groupby(['insult'])['insult'].count())
print ('')
print (df_train.groupby(['identity_hate'])['identity_hate'].count())

#  Threat has 1,904 records
#  Severe toxic has 2,423 records
#  Toxic has 17,873
#  Obscene has 9,831
#  Insult has 9,213
#  Identity hate has 2,419


threat  severe_toxic
0       1               1983
1       0               1464
        1                440
Name: threat, dtype: int64

Total records:  16414
severe_toxic
0    13991
1     2423
Name: severe_toxic, dtype: int64

threat
0    14510
1     1904
Name: threat, dtype: int64

toxic
0    10219
1     6195
Name: toxic, dtype: int64

obscene
0    12202
1     4212
Name: obscene, dtype: int64

insult
0    12306
1     4108
Name: insult, dtype: int64

identity_hate
0    13995
1     2419
Name: identity_hate, dtype: int64


In [32]:
add_back_translation = True
add_toxigen_for_identity_hate = True

def summarize_df():
  print('threat:  ' + str(len(df[(df['flag']==True) & (df['threat']==1)])))
  print('identity_hate:  ' + str(len(df[(df['flag']==True) & (df['identity_hate']==1)])))
  print('severe_toxic:  ' + str(len(df[(df['flag']==True) & (df['severe_toxic']==1)])))
  print('insult:  ' + str(len(df[(df['flag']==True) & (df['insult']==1)])))
  print('obscene:  ' + str(len(df[(df['flag']==True) & (df['obscene']==1)])))
  print('toxic:  ' + str(len(df[(df['flag']==True) & (df['toxic']==1)])))


if add_back_translation and add_toxigen_for_identity_hate:
  df = df_train.copy()

  #  The tag with the lowest count is Threat, so lets flag all of them to keep
  df['flag'] = np.where(df['threat']== 1, True, False)
  summarize_df()


threat:  1904
identity_hate:  388
severe_toxic:  440
insult:  1224
obscene:  1204
toxic:  1788


In [35]:

#  Drops a few records with NA.  Less than 5.  Not significant.
print ('Len after importing:  ' + str(len(df_train)))
df_train = df_train[df_train['toxic'].notna()]
print ('Len after dropping NAs:  ' + str(len(df_train)))

df_train, df_val = train_test_split(df_train,
                                  stratify=df_train[['toxic', 'severe_toxic', 'obscene', 'threat', 'insult', 'identity_hate']],
                                  test_size=0.2,
                                  random_state=1)

Len after importing:  16414
Len after dropping NAs:  16414


In [36]:
print (df_train.shape)
display (df_train.head(5))
print ('Toxic:  ' + str(len(df_train[df_train.toxic == 1])))
print ('Severe Toxic:  ' + str(len(df_train[df_train.severe_toxic == 1])))
print ('Obscene:  ' + str(len(df_train[df_train.obscene == 1])))
print ('Threat:  ' + str(len(df_train[df_train.threat == 1])))
print ('Insult:  ' + str(len(df_train[df_train.insult == 1])))
print ('Identity Hate:  ' + str(len(df_train[df_train.identity_hate == 1])))



print ('\n\n\n')
print (df_val.shape)
display (df_val.head(5))
print ('Toxic:  ' + str(len(df_val[df_val.toxic == 1])))
print ('Severe Toxic:  ' + str(len(df_val[df_val.severe_toxic == 1])))
print ('Obscene:  ' + str(len(df_val[df_val.obscene == 1])))
print ('Threat:  ' + str(len(df_val[df_val.threat == 1])))
print ('Insult:  ' + str(len(df_val[df_val.insult == 1])))
print ('Identity Hate:  ' + str(len(df_val[df_val.identity_hate == 1])))


print ('\n\n\nTest Stratification:')
print ('Toxic:  ', end='')
print (len(df_val[df_val.toxic == 1]) / len(df_train[df_train.toxic == 1]))

print ('Severe Toxic:  ', end='')
print (len(df_val[df_val.severe_toxic == 1]) / len(df_train[df_train.severe_toxic == 1]))

print ('Obscene:  ', end='')
print (len(df_val[df_val.obscene == 1]) / len(df_train[df_train.obscene == 1]))

print ('Threat:  ', end='')
print (len(df_val[df_val.threat == 1]) / len(df_train[df_train.threat == 1]))

print ('Insult:  ', end='')
print (len(df_val[df_val.insult == 1]) / len(df_train[df_train.insult == 1]))

print ('Identity Hate:  ', end='')
print (len(df_val[df_val.identity_hate == 1]) / len(df_train[df_train.identity_hate == 1]))



(13131, 8)


id                                       comment_text  \
5303  280d91309096c5c6  lead paragraph the lead does not refers that t...   
7394  d665a0a940e907de  the origins of schizophrenia published 2012 by...   
4661  0f0cfa0ac495b826  tmans comments   having had a third opnion by ...   
1541  a9e692d5eb8a2496  i would be happy to accept some authority for ...   
15    a9392d154b54b01c  lewes the county town of sussexs is not quite ...   

      toxic  severe_toxic  obscene  threat  insult  identity_hate  
5303      0             0        0       0       0              0  
7394      0             0        0       0       0              0  
4661      0             0        0       0       0              0  
1541      0             0        0       0       0              0  
15        0             0        0       0       0              0

Toxic:  4957
Severe Toxic:  1940
Obscene:  3368
Threat:  1522
Insult:  3286
Identity Hate:  1935




(3283, 8)


id                                       comment_text  \
13452           toxigen   when it comes down to it latinos are mostly g...   
13668  d0a55e6ffaa62b3d            kiss my ass  eat shit and die 680119139   
6629   66b8924c284d139b  sorry but i know nothing of munich and honestl...   
2024   8da9b88e0c7218c0   please unblock me i am serious about never ed...   
13632  38cd26dc19b19b48              holy shit shut the f up you motherf c   

       toxic  severe_toxic  obscene  threat  insult  identity_hate  
13452      1             0        0       0       0              1  
13668      1             1        1       1       1              0  
6629       0             0        0       0       0              0  
2024       0             0        0       0       0              0  
13632      1             1        1       0       1              0

Toxic:  1238
Severe Toxic:  483
Obscene:  844
Threat:  382
Insult:  822
Identity Hate:  484



Test Stratification:
Toxic:  0.24974783134960662
Severe Toxic:  0.24896907216494846
Obscene:  0.2505938242280285
Threat:  0.25098554533508544
Insult:  0.25015216068167984
Identity Hate:  0.2501291989664083


In [37]:

df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)



In [38]:
df_train.to_csv('train_data.csv', index=False)
df_val.to_csv('validation_data.csv', index=False)
#df_test.to_csv('test_data.csv', index=False)

In [39]:

files.download('train_data.csv')
files.download('validation_data.csv')
#files.download('test_data.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Then upload to AWS

Then check that files can be read from AWS

In [40]:
df_aws_train = pd.read_csv("https://adamhyman-public.s3.amazonaws.com/w266/for_modeling/train_data.csv")
df_aws_validation = pd.read_csv("https://adamhyman-public.s3.amazonaws.com/w266/for_modeling/validation_data.csv")
df_aws_test = pd.read_csv("https://adamhyman-public.s3.amazonaws.com/w266/for_modeling/test_data.csv")

df_aws_train_balanced = pd.read_csv('https://adamhyman-public.s3.amazonaws.com/w266/for_modeling/augmented_and_balanced/train_data_balanced.csv')
df_aws_validation_balanced = pd.read_csv('https://adamhyman-public.s3.amazonaws.com/w266/for_modeling/augmented_and_balanced/validation_data_balanced.csv')

display (df_aws_train.head(3))
display (df_aws_validation.head(3))
display (df_aws_test.head(3))

display (df_aws_train_balanced.head(3))
display (df_aws_validation_balanced.head(3))

print (df_aws_train.shape)
print (df_aws_validation.shape)
print (df_aws_test.shape)

id                                       comment_text  toxic  \
0  1f248a9d4e6a747e                   unsigned comment by userjohn4890      0   
1  4cb60854285a70b8   how can you block me when you are just an editor      0   
2  4462a926cf64b036    berwickshire i have added the category berwi...      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       0              0

id                                       comment_text  toxic  \
0  d6d594838a3870c0     dyk suggestion olvir hnufa   hi brian  i en...      0   
1  c7f79f4af1e0aad6  junior barranquilla  wtf how did you come up w...      1   
2  d4495c7b6fda20fc  and whats the point of editing a file if your ...      1   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       1              0

id                                       comment_text  toxic  \
0  0001ea8717f6de06  thank you for understanding i think very highl...      0   
1  000247e83dcc1211                     dear god this site is horrible      0   
2  0002f87b16116a7f   somebody will invariably try to add religion ...      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       0              0

id                                       comment_text  toxic  \
0  280d91309096c5c6  lead paragraph the lead does not refers that t...      0   
1  d665a0a940e907de  the origins of schizophrenia published 2012 by...      0   
2  0f0cfa0ac495b826  tmans comments   having had a third opnion by ...      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       0              0

id                                       comment_text  toxic  \
0           toxigen   when it comes down to it latinos are mostly g...      1   
1  d0a55e6ffaa62b3d            kiss my ass  eat shit and die 680119139      1   
2  66b8924c284d139b  sorry but i know nothing of munich and honestl...      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              1  
1             1        1       1       1              0  
2             0        0       0       0              0

(127655, 8)
(31914, 8)
(63972, 8)
